pkg/scheduler/scheduler.go

```diff
@@ -0,0 +1,336 @@
+		//
+		// Note that the key is added directly to the queue without having to wait for any rate limiter's
+		// approval. This is necessary as requeues, requested by the scheduler, occur when the scheduler
+		// is certain that more scheduling work needs to be done but it cannot be completed in
+		// one cycle (e.g., a plugin sets up a per-cycle batch limit, and consequently the scheduler must
+		// finish the scheduling in multiple cycles); in such cases, rate limiter should not add
+		// any delay to the requeues.
+		s.queue.Add(crpName)
+	}
+}
+
+// Run starts the scheduler.
+//
+// Note that this is a blocking call. It will only return when the context is cancelled.
+func (s *Scheduler) Run(ctx context.Context) {
+	klog.V(2).InfoS("Starting the scheduler")
+	defer func() {
+		klog.V(2).InfoS("Stopping the scheduler")
+	}()
+
+	// Starting the scheduling queue.
+	s.queue.Run()
+
+	// Run scheduleOnce forever.
+	//
+	// The loop starts in a dedicated goroutine; it exits when the context is canceled.
+	go wait.UntilWithContext(ctx, s.scheduleOnce, 0)
+
+	// Wait for the context to be canceled.
+	<-ctx.Done()
+
+	// Stopping the scheduling queue; drain if necessary.
+	//
+	// Note that if a scheduling cycle is in progress; this will only return when the
+	// cycle completes.
+	s.queue.CloseWithDrain()
+}
+
+// cleanUpAllBindingsFor cleans up all bindings derived from a CRP.
+func (s *Scheduler) cleanUpAllBindingsFor(ctx context.Context, crp *fleetv1beta1.ClusterResourcePlacement) error {
+	crpRef := klog.KObj(crp)
+
+	// List all bindings derived from the CRP.
+	//
+	// Note that the listing is performed using the uncached client; this is to ensure that all related
+	// bindings can be found, even if they have not been synced to the cache yet.
+	bindingList := &fleetv1beta1.ClusterResourceBindingList{}
+	listOptions := &client.ListOptions{
+		LabelSelector: labels.SelectorFromSet(labels.Set{fleetv1beta1.CRPTrackingLabel: crp.Name}),
+	}
+	// TO-DO (chenyu1): this is a very expensive op; explore options for optimization.
+	if err := s.uncachedReader.List(ctx, bindingList, listOptions); err != nil {
+		klog.ErrorS(err, "Failed to list all bindings", "ClusterResourcePlacement", crpRef)
+		return controller.NewAPIServerError(false, err)
+	}
+
+	// Remove the scheduler cleanup finalizer from all the bindings, and delete them.
+	//
+	// Note that once a CRP has been marked for deletion, it will no longer enter the scheduling cycle,
+	// so any cleanup finalizer has to be removed here.
+	//
+	// Also note that for deleted CRPs, derived bindings are deleted right away by the scheduler;
+	// the scheduler no longer marks them as deleting and waits for another controller to actually
+	// run the deletion.
+	for idx := range bindingList.Items {
+		binding := bindingList.Items[idx]
+		// Delete the binding if it has not been marked for deletion yet.
+		if binding.DeletionTimestamp == nil {
+			if err := s.client.Delete(ctx, &binding); err != nil {
```

Comment:
ignore the notfounderror?

in case the rollout controller just deletes the binding

Reply:
Done.